__tests__/focus-prism.test.tsx

```diff
@@ -65,3 +67,53 @@ it('atoms that focus on no values are not updated', async () => {
   await findByText('count:')
   await findByText('bigAtom: {}')
 })
+
+type BillingData = {
+  id: string
+}
+
+type CustomerData = {
+  id: string
+  billing: BillingData[]
+  someDynamicData?: string
+}
+
+it('typescript should work well with nested arrays containing optional values', async () => {
+  const customerListAtom = atom<CustomerData[]>([])
+
+  const foundCustomerAtom = focusAtom(customerListAtom, (optic) =>
+    optic.find((el) => el.id === 'some-invalid-id')
+  )
+
+  const derivedAtom = focusAtom(foundCustomerAtom, (optic) => {
+    const result = optic
+      .valueOr({ billing: [] } as unknown as CustomerData)
+      .prop('billing')
+      .find((el) => el.id === 'some-invalid-id')
+
+    return result
+  })
+
+  expectTypeOf(derivedAtom).toMatchTypeOf<
+    WritableAtom<BillingData | undefined, SetStateAction<BillingData>, void>
+  >()
+})
+
+it('should work with promise based atoms with "undefined" value', async () => {
+  const customerBaseAtom = atom<CustomerData | undefined>(undefined)
+
+  const asyncCustomerDataAtom = atom(
+    async (get) => get(customerBaseAtom),
+    (_, set, nextValue: CustomerData) => {
+      set(customerBaseAtom, nextValue)
+    }
+  )
+
+  const focusedPromiseAtom = focusAtom(asyncCustomerDataAtom, (optic) =>
+    optic.optional()
+  )
+
+  expectTypeOf(focusedPromiseAtom).toMatchTypeOf<
+    WritableAtom<Promise<string | undefined>, SetStateAction<CustomerData>>
+  >
```

Comment:
@dai-shi shouldn't this fail too? 🤔 

The Type would be narrowed to `WritableAtom<CustomerData | undefined, SetStateAction<CustomerData>, void>`, no?

Reply:
yeah, something was totally wrong. 😓